libopenage/renderer/glyph_atlas.cpp

```diff
@@ -0,0 +1,153 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#include "glyph_atlas.h"
+
+#include <functional>
+#include <cstring>
+#include <epoxy/gl.h>
+
+#include "../error/error.h"
+
+namespace openage {
+namespace renderer {
+
+GlyphAtlas::Shelf::Shelf(int y_position, int width, int height)
+	:
+	y_position{y_position},
+	width{width},
+	height{height},
+	remaining_width{width} {
+	// Empty
+}
+
+bool GlyphAtlas::Shelf::check_fits(int required_width, int required_height) {
+	if (required_width > this->remaining_width || required_height > this->height) {
+		return false;
+	}
+	return true;
+}
+
+int GlyphAtlas::Shelf::reserve(int required_width, int/* required_height*/) {
+	int x_position = this->width - this->remaining_width;
+	this->remaining_width -= required_width;
+	return x_position;
+}
+
+GlyphAtlas::GlyphAtlas(int width, int height)
+	:
+	width{width},
+	height{height},
+	is_dirty{false},
+	dirty_area{width, height, 0, 0},
+	texture_id{0} {
+
+	this->buffer = std::unique_ptr<unsigned char>(new unsigned char[this->width * this->height]);
+	memset(this->buffer.get(), 0, this->width * this->height);
+
+	glGenTextures(1, &this->texture_id);
+	glBindTexture(GL_TEXTURE_2D, this->texture_id);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+
+	glTexImage2D(GL_TEXTURE_2D, 0, GL_RED, this->width, this->height, 0, GL_RED, GL_UNSIGNED_BYTE, this->buffer.get());
```

Comment:
Not necessarily. `GlyphAtlas` itself can be made independent of backend renderer. I can remove the `gl` calls inside the atlas and provide an api (let's say `get_buffer`). `TextRenderer` (or any other manager) can use this api to flush the buffer to texture.

This can be done once renderer is merged :D